### PR TITLE
Fix error when SubscriptionIntraProcess is being executed more than once by the MultiThreadedExecutor

### PR DESCRIPTION
**Pull request by ivanpauno**
This ensures that executing is a noop if there's nothing to be executed.

Alternative to https://github.com/ros2/rclcpp/pull/1241.

### CODE REVIEW

File: rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp
Comment (audrow) [inline rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp:160]: I think that you want a `lock_guard` before this statement. I can imagine a segmentation fault as follows:
* Thread 1: `!buffer_->has_data()` evaluates to `false` so the thread skips over the if-statement
* Thread 2: consumes the data in `buffer_`
* Thread 1: Tries to consume the data that has already been consumed.

Comment (ivanpauno) [reply rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp:160]: Yeap, thanks for noting the issue!

Comment (ivanpauno) [reply rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp]: Solved in https://github.com/ros2/rclcpp/pull/1275/commits/3d0f8df8a86d93d362dc79cd7aaa95dcd9f9031b.

**Comment by ivanpauno** [comment]
> I like how minimal this solution is. As you mention that this is just a fix for SubscriptionIntraProcess, the same kind of fix should be applied to all classes that inherit from Waitable

All the other Waitables that https://github.com/ros2/rclcpp/pull/1241 is modifying still have the same issue after and before the PR (for QOSEventHandler, action clients and services, #1241 isn't fixing the problem).
That's a big reason why I don't think #1241 is correct.

I can apply a fix to action clients and services, doing the same that subscriptions do: ignore when taking data fails.
I would prefer to do that in a separate PR.

For `QOSEventHandler`, the only possible fix is to add the concept of a Waitable that is mutually exclusive with itself (which can be implemented by creating a callback group and adding only that single Waitable to it).
Actually, I think this approach is nice for timers too.

**Comment by ivanpauno** [comment]
> All the other Waitables that #1241 is modifying still have the same issue after and before the PR (for QOSEventHandler, action clients and services, #1241 isn't fixing the problem).

Correction: #1241 does fix the issue for action clients and services.

File: rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp
Comment (sloretz) [inline rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp:89]: @ivanpauno @audrow This looks like it breaks ABI. Is it possible to implement this in a way that can be back ported to Foxy? One of the Foxy CI failures looks like it might be an instance of #1212 

http://build.ros2.org/job/Fci__nightly-fastrtps_ubuntu_focal_amd64/193/testReport/junit/(root)/projectroot/gtest_multithreaded__rmw_fastrtps_cpp/

Comment (ivanpauno) [reply rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp:89]: I can avoid deleting that line, but the patch will still break ABI because of the addition of:
https://github.com/ros2/rclcpp/blob/3d0f8df8a86d93d362dc79cd7aaa95dcd9f9031b/rclcpp/include/rclcpp/timer.hpp#L260

I don't see how to avoid breaking ABI.

File: rclcpp/include/rclcpp/timer.hpp
Comment (sloretz) [inline rclcpp/include/rclcpp/timer.hpp:207]: Is this related to #1212? It doesn't seem related to the intra process ring buffer being taken from twice. Could this be it's own pull request? Would that then allow backporting the #1212 fix to Foxy?

Comment (ivanpauno) [reply rclcpp/include/rclcpp/timer.hpp:207]: > Could this be it's own pull request?

Yes, I actually should open an issue describing the problem.
I originally solved both here because they are related.

> Would that then allow backporting the #1212 fix to Foxy?

:+1:

Comment (ivanpauno) [reply rclcpp/include/rclcpp/timer.hpp]: done, sorry for the delay

**Review by fujitatomoya — APPROVED** [review]
lgtm

**Review by audrow — APPROVED** [review]

**Comment by ivanpauno** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12864/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7819/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10575/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12807/)

**Comment by wjwwood** [comment]
I generally do not like this approach, as I said here:

https://github.com/ros2/rclcpp/pull/1241#issuecomment-721421457

But I'll leave it to the other `rclcpp` maintainers to decide, since I don't have time to help architect a better solution.

**Pull request closed without merging**
